### ISSUE TITLE
Hardcoded CXX Compiler (Clang/GCC 5.0 abi incompatibility)

### ISSUE THREAD

**Issue by uaftab** [pfpsim/pfpsim-installer#2]
The PFPSIM Installer by default uses `clang++` for compiling packages from source and the system default c++ compiler.

In the release of `GCC 5.1 libstdc++`, a new library abi was introduced for conformance with the C++11 Standard. For backwards compatibility of code linking to `libstdc++` the gcc developers opted to keep the library's soname to be same., instead new implementations are achieved by inlining namespaces so they have different names for linking all new c++11 code is linked via with the `std::__cx11::` tag in the symbols namespace which is controlled by the `_GLIBCXX_USE_CXX11_ABI` macro whether a library is compiled with the tag or not. 

> https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

ABI incompatibility warning can be enabled in gcc with the `-Wabi-tag` option.

The dual abi is not yet supported by clang/llvm upstream everywhere and is listed as an active bug [here](https://llvm.org/bugs/show_bug.cgi?id=23529). ArchLinux is the only dist that seems to be patched as of 16/06/2016 [[Ref1](https://llvm.org/bugs/show_bug.cgi?id=23529#c37)] [[Ref2](https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/llvm#n81)]. This will cause linking errors if some packages are compiled using clang which if a fad project being compiled using gcc will try to link to. Also clang will not be able to link to packages installed using a package manager (apt-get, aptitude, dnf) as those packages are probably also compiled with the new abi. 
### Fix:

Installer now uses the default compiler available on your system and compiles all packages using that. 

| Distribution | Status GCC | Status Clang | System Default Abi Version | Comments |
| :-: | :-: | :-: | :-: | :-: |
| Debian 8.x, 7.x | Stable - :white_check_mark: | Stable - :white_check_mark: | 4.9 | System can be upgraded to GCC5.0 but issue report is open [#797917](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=797917) |
| ArchLinux | Stable - :white_check_mark: | Stable - :white_check_mark: | 5.x | ArchLinux has patched its upstream [LLVM](https://llvm.org/bugs/show_bug.cgi?id=23529#c37) |
| Mint 17.3 | Stable - :white_check_mark: | Stable - :white_check_mark: | 4.9 | - |
| Ubuntu 14.04 LTS | Stable - :white_check_mark: | Stable - :white_check_mark: | 4.9 | - |
| Ubuntu 15.x | Stable - :white_check_mark: | Active BUG - :x: | 5.x | [LaunchPad Bug #1488254](https://bugs.launchpad.net/ubuntu/+source/llvm-toolchain-3.6/+bug/1488254) |
| Ubuntu 16.04 LTS | Stable - :white_check_mark: | Active BUG - :x: | 5.x | [LaunchPad Bug #1488254](https://bugs.launchpad.net/ubuntu/+source/llvm-toolchain-3.6/+bug/1488254) |

**_Please NOTE if your system's default compiler is clang/llvm and the libstdc++ abi is 5.0 and if the abi5.0 bug hasn't been patched for llvm by your distribution provider it will FAIL TO COMPILE._**  
### For CLANG users on affected dist:

Fix1: 
 *\* Downgrade everything to the 4.9 abi**
- Remove GCC-5
- Downgrade the gcc suite (gcc,g++,libstdc++-dev) to 4.9
- Compile all apt-get dependencies with g++/gcc-4.9 for their system for pfpsim and pfp-p4.
  - libboost, libssl, etc. 
- Compile & Install the whole pfpsim suite with 4.9

Fix2: 
- Use GCC instead clang

Fix3:
- Switch to a different distribution like Debian or ArchLinux _[the bleeding-edge dist](https://www.archlinux.org/about/)_
### TL; DR

All Packages if using clang need to be compiled with libstdc++ abi version 4.9 or lower, till this [bug is resolved upstream](https://llvm.org/bugs/show_bug.cgi?id=23529).
For GCC it works.
